### ISSUE TITLE
OSM XML output produces 0 byte file

### ISSUE THREAD

**Issue by arminus** [clarisma/geodesk#44]
First of all, thanks a lot for providing this, I'm super interested!

In my first test, I loaded the entire current planet:
```
gol build planet ../osm/data/planet-20221108.pbf
Building planet.gol from ../osm/data/planet-20221108.pbf using default settings...
12:24:13 Minimum instance count is now 3
Analyzed ../osm/data/planet-20221108.pbf in 3m 31s
12:42:18 Rejected way/1096316478 because of invalid node count (1)          
12:42:22 Rejected way/1110722485 because of invalid node count (1)          
12:42:22 Rejected way/1111185820 because of invalid node count (1)          
12:42:22 Rejected way/1110308529 because of invalid node count (1)          
Sorted 8,029,123,737 nodes / 899,849,990 ways / 10,385,291 relations in 15m 34s
12:54:29 Could not calculate quad for relation/14515115 [processing 6/53/24]
Validated 29593 tiles in 12m 11s
Compiled 29593 tiles in 17m 53s
13:12:52 relation/14245409 has not been exported by tile 000002, can't resolve at 000000/00000160
Linked 29593 tiles in 58s 292ms
Built planet.gol in 50m 20s
```

First query works ok:
```
gol query ./planet.gol "nwr[via_ferrata_scale]" -f=geojson > via_ferrata_scale.geojson

Retrieved 3,790 features in 3s 489ms
```

But running this with xml as output type produces a 0 byte file:
```
gol query ./planet.gol "nwr[via_ferrata_scale]" -f=xml > via_ferrata_scale.osm

Retrieved 3,790 features in 1s 786ms
❯ ll via_ferrata_scale.osm
-rw-r--r-- 1 arminus arminus 0 Nov  9 13:48 via_ferrata_scale.osm
```

**Comment by clarisma** [comment]
Thanks for the feedback!
We ended up deferring OSM-XML output to Version 0.2, but I forgot to mark it in the documentation.
Sorry about that, I'll update the docs now.

**Comment by clarisma** [comment]
Thanks! Looks like `geojson` output is not encoding `GeometryCollection` objects correctly.
I'll log this as a separate bug for the GOL tool.
Most likely, we'll ship XML output as part of the next maintenance release (0.1.1), along with a fix for GeoJSON, so you won't have to wait until 0.2.

**Comment by clarisma** [comment]
By the way, if you're looking for even faster queries, try to include a tag whose key is indexed as part of your query, e.g.: `r[route][via_ferrata_scale]`. This causes the engine to use the index for `route` relations.
You can see which keys are indexed by running `gol info planet`.

**Comment by arminus** [comment]
I don't want to misuse this ticket as discussion media, but as far as performance in my data ingestion pipeline is concerned, initial staging is a key. Currently, it takes osmupdate approx 45 minutes to update my o5m planet file (run daily), then I have various osmfilter steps on the planet file, each of which takes about 25 minutes.

So with geodesk, this would add an initial ~ 50 minutes to rebuild the planet gol file and "only" save time later. Are there any plans to incrementally update an existing gol file from the standard osc files?

**Comment by clarisma** [comment]
Yes, incremental updates are on our roadmap. This feature is tentatively slated for Version 0.3 (We don't have an exact a timeline yet).

GOL files are designed to be patchable. In theory, we could accommodate minutely updates, but our initial design aims for daily updates (The update-processing software would require greater complexity to do its job in 60 seconds).

Instead of directly applying `.osc` change files to the GOL, a separate program first pre-processes each `.osc` file into a set of *patch files* (one for each tile in which features have changed) and a *version manifest* (a small file that indicates which tiles have changed). 

Ideally, this would run on a central server, so downstream users can skip the `build` import (which is comparatively resource-intensive) and just download tiles and patches.

As a data consumer, your workflow would then look like this:

Instead of 

> `build planet <osm-pbf-file>`

you would use 

> `load planet <tileset-url> [ -a=<area> | -b=<bbox> ]`
 
to download the current version of the planet (or only the regions you require). The compressed tiles are similar in size to data in osm-pbf format. The tiles are simply unzipped and written into the GOL, so this step happens essentially at download speed. 

Using

> `update planet <tileset-url>`
 
you can then periodically update your GOL to the most current state of the data. This should only take seconds if your GOL is reasonably current (again, basically whatever time is required to download the patches).   

Updating happens as a two-step process: First, the client-side engine downloads the version manifest (which indicates the tiles that changed) and marks those tiles in the user's GOL file as *stale*. It then downloads patch files that contain the changes and applies them to the tiles to bring them to the current version. 

Updating can be *eager* (all stale tiles are immediately brought current by the `update` command) or *lazy* (patches are downloaded and applied to stale tiles only once they are actually required by a query).

Our work is still early-stage on this feature. I'll post it as a separate issue to make it easier to gather feedback on how the details of it should evolve.

**Comment by arminus** [comment]
Thanks for the quick fix. At first glance, the files are looking good. But I need a bit more time to feed them into a test pipeline, probably next week.

I'm also using osmium derive-changes to produce diffs between the current and previous extracts, that's also something I can only test after a later planet update.

**Comment by arminus** [comment]
Meanwhile, I've tested this with osmium derive-changes for 2 osm files produced by gol query "nwr[via_ferrata_scale]" -f=xml and D/A/CH dumps/gol files which where 4 days apart.

Bad news there, the produced osc file is ~ 3.5 MB, while the osc files generated from the OSM files as produced by osmfilter is a reasonable 8.4 KB. Haven't investigated what preconditions osmium has on the OSM files, but apparently, those produced by gol query don't work for that usecase.

**Comment by clarisma** [comment]
If I understand correctly, you are getting an `.osc` that is much larger than you were expecting.

I see what's going on here...

GOLs don't store IDs for nodes unless they are tagged or are part of relations. When producing XML output, the GOL tool re-creates IDs for these "anonymous" nodes, so the XML file has referential integrity internally. These "synthetic" IDs don't remain stable from one XML output to the next (even if the query results are the same) -- that's why `derive-changes` reports a far greater number of changes.

Two ways we could address this:

A. Deliver objects in XML output in sorted order; this makes the output deterministic and should result in less variance of the synthetic node IDs.

B. Provide an option during `build` to keep all node IDs. This way, XML output will contain the true node IDs (though this would increase the size of a GOL by about 20%).

A is an easy fix (related to Issue #45); B is something we've contemplated, but will require additional development & testing.

I've also opened (A) as a separate issue for the GOL tool: clarisma/gol-tool#64

**Comment by arminus** [comment]
> GOLs don't store IDs for nodes unless they are tagged or are part of relations.

So you're saying that all untagged nodes of ways are stored with synthetic IDs? My diff usecase might be kind of special, but aren't synthetic IDs as retrieved by gol query a general limitation for other downstream processing? (if your objective is to be a replacement for osmfilter, overpass or planet database queries)

**Comment by clarisma** [comment]
If a node is untagged (and does not participate in a relation) -- i.e. its sole purpose is to define the geometry of a way -- it is considered to have no identity at all. During import, GeoDesk does not store such nodes as individual features. Instead of a list of node ids, each way stores a list of coordinates (This eliminates the need to look up each node, and therefore makes geometry creation much faster).

If you use the Java API and call `Way.nodes()`, you will get a series of `Node` objects, but only the *feature nodes* (e.g. a crosswalk or speed bump along a road) will have an ID. All other nodes have an ID of `0` (they merely carry their coordinates).

When query results are exported as OSM-XML, the GOL Tool assigns a synthetic ID to each anonymous node. This ensures referential integrity among ways and nodes within the XML file (the nodes of two connected ways will have the same ID). So, within the same XML file: a) all anonymous nodes with the same coordinates will have the same ID, and b) synthetic IDs will not clash with the "real" ID of a feature node within the same file.

This way, downstream programs can reassemble the geometries of the exported features from a single XML file. However, merging multiple such XML exports will likely cause the synthetic node IDs to clash (Referential integrity is only guaranteed within the same file). Also, you cannot use XML exports for editing (though you can visualize them in JOSM).  

What specifically are your downstream processes doing with these XML outputs?

**Comment by arminus** [comment]
Thanks for the detailed explanation.

> What specifically are your downstream processes doing with these XML outputs?

Being able to diff with previous OSM-XMLs for sure. And I do in fact merge OSM-XMLs from different osmfilter queries into one for input to a routing engine, which as I understand it would presently also not work with your OSM-XMLs.

But don't let me influence your overall design decisions too much, I'm just running a small site for via_ferratas and cross country skiing / skitouring. I haven't thought about further usecases for geodesk, at this point, using it would "only" speed up my data input pipeline.

I do have a feeling though, that these synthetic IDs might cause issues with other OSM-XML consumers in general. Replacing a heavy-weight full planet DB with a .gol file is after all a generally appealing usecase - apart from what I'm doing. 

So your option B above might seem prudent.

**Comment by clarisma** [comment]
I created a roadmap issue for an option to store the IDs of all nodes (#57).

If a routing engine assumes "same location = same node", then the synthetic node IDs should not create issues. On the other hand, if it uses "same location, different ID = different node", then there will be disconnected segments, and routing will fail.

You could also collect all features of interest using a single query (multiple selectors can be listed, separated by commas).

Btw, what's the url of your site? Sounds interesting!

**Comment by arminus** [comment]
> You could also collect all features of interest using a single query (multiple selectors can be listed, separated by commas).

Interesting, but that probably won't help because I'm doing some other processing on the OSM-XMLs before I merge them back together. I'll have a deeper look eventually, though.

> Btw, what's the url of your site? Sounds interesting!

https://www.xctrails.org (the map app can be switched to English, the landing page is only German)
